### PR TITLE
createItemType Via Lua

### PR DESCRIPTION
**Pull request by MillhioreBT**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

**Issues addressed:** Inspired by this PR #3010

### Some peculiarities:
* Items can be created through lua if you are in the Rev Scripts environment
* However its charge and recharge are separate
* Can't reload with `/reload scripts` instead use `/reload items` as commonly done with xml

### Note:
* I have no reason to say that this is the final version, improvements in the code are probably needed
* Although I have been testing these changes for almost 2 months, it has not been peer-reviewed yet, so if you notice something weird and want to help me improve this, your suggestions are welcome.
* Note that the `items.lua` file is by convention, but you can create all the luas files you want in the `data/items` folder so that you can sort sets of items by file as you wish.

How it looks:
```lua
Game.createItemType(26378):register({
   name = "gold converter",
   article = "a",
   weight = 1800,
   duration = 604800,
   charges = 500,
   showcharges = true,
   showduration = true
})
```

### CODE REVIEW

File: src/items.h [hunk 1]
```diff
@@ -312,6 +322,8 @@ class ItemType
 		std::unique_ptr<Abilities> abilities;
 		std::unique_ptr<ConditionDamage> conditionDamage;
 
+		std::map<std::string, bool> hasEdited;
```
Comment (MillhioreBT) [inline src/items.h:325]: Here I will use a `unordered_map`, I am hoping to find more things to improve the code

File: data/lib/compat/compat.lua [hunk 2]
```diff
@@ -1398,14 +1399,16 @@ function createFunctions(class)
...
-			if not (rawget(class, get) and rawget(class, set)) then
-				table.insert(temp, {set, setFunc, get, getFunc})
+			local is = "is" .. str
+			if not (rawget(class, get) and rawget(class, set) and rawget(class, is)) then
+				table.insert(temp, {set, setFunc, get, getFunc, is, getFunc})
 			end
 		end
 	end
-	for _, func in ipairs(temp) do
+	for _,func in ipairs(temp) do
```
Comment (EPuncker) [inline data/lib/compat/compat.lua:1408]:
```suggestion
	for _, func in ipairs(temp) do
```

**Comment by DSpeichert** [comment]
If you're constraining this to just `data/items/` then isn't it going against modularity of Revscripts as mentioned somewhere else by @EvilHero90 ?

**Comment by EvilHero90** [comment]
For now we can't really get around having it extra like we did with monsters because of the loading order (might throw errors and such if we load items from lua later than we do at xml), at current state as we approach further into lua handling stuff this is already a good enhancement and if we every going to rely on lua only we can start re arranging loading order and such so we have everything in one place.

Resolve the conflicts and then we should finaly merge it.

File: src/luascript.cpp [hunk 3]
```diff
@@ -11431,764 +11652,1901 @@ int LuaScriptInterface::luaItemTypeCreate(lua_State* L)
...
+				abilities.absorbPercent[combatTypeToIndex(COMBAT_EARTHDAMAGE)] += value;
+				abilities.absorbPercent[combatTypeToIndex(COMBAT_ICEDAMAGE)] += value;
+				abilities.absorbPercent[combatTypeToIndex(COMBAT_HOLYDAMAGE)] += value;
+				abilities.absorbPercent[combatTypeToIndex(COMBAT_DEATHDAMAGE)] += value;
+				pushBoolean(L, true);
+				return 1;
+			} else if (abilitieName == "absorbpercentall") {
+				int16_t value = getNumber<int16_t>(L, 3);
+				for (uint8_t i = COMBAT_NONE; i <= COMBAT_COUNT; i++) {
+					abilities.absorbPercent[i] += value;
```
Comment (EvilHero90) [inline src/luascript.cpp:13461]:
```suggestion
					abilities.absorbPercent[indexToCombatType(i)] += value;
```

File: src/luascript.cpp [hunk 4]
```diff
@@ -11431,764 +11652,1901 @@ int LuaScriptInterface::luaItemTypeCreate(lua_State* L)
...
+			// FieldAbsorbPercents
+			auto fieldAbsorbPercentVal = AbilitieFieldAbsorbPercentMap.find(abilitieName);
+			if (fieldAbsorbPercentVal != AbilitieFieldAbsorbPercentMap.end()) {
+				abilities.fieldAbsorbPercent[combatTypeToIndex(fieldAbsorbPercentVal->second)] += getNumber<int16_t>(L, 3);
+				pushBoolean(L, true);
+				return 1;
+			} else if (abilitieName == "fieldabsorbpercentall") {
+				int16_t value = getNumber<int16_t>(L, 3);
+				for (uint8_t i = COMBAT_NONE; i <= COMBAT_COUNT; i++) {
+					abilities.fieldAbsorbPercent[i] += value;
```
Comment (EvilHero90) [inline src/luascript.cpp:13476]:
```suggestion
					abilities.fieldAbsorbPercent[indexToCombatType(i)] += value;
```

**Review by EvilHero90 — CHANGES_REQUESTED** [review]
tons of formatting issues in `luascript.cpp` which need to be resolved but other then that it looks fine.

**Pull request closed without merging**